### PR TITLE
[inductor] Triton mm codegen uses `make_block_ptr`

### LINKED ISSUE #109420: Supporting Block_Ptrs in inductor code gen
### Summary
More curiosity and a tracking issue for me or someone else to pick up.
Triton has introduced Block_pointer: https://triton-lang.org/main/getting-started/tutorials/08-experimental-block-pointer.html

Personally I think these are easier to work with/reason about than raw strides. I could be missing something fundamental to inductors loop level IR but would be interesting to see what would be needed to support usage in this for example:
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm.py#L31-L91 

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### PR DESCRIPTION
**Pull request by jon-chuang**
Partially fixes: https://github.com/pytorch/pytorch/issues/109420

No perf degr (on RTX 4070 Laptop Ada Lovelace):

Before:
image

After:
image

The reason that the curve is not equal to cublas/triton for smaller M/N is because the triton autotune configs are not tuned for commercial GPUs. See: https://github.com/pytorch/pytorch/issues/109489

`uint4x2_mixed_mm` is skipped due to complicated use of the ptrs to do bitwise hacks.

CC: @jansel @drisspg 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by drisspg** [comment]
@eellison we should probably run a inductor perf run on this I imagine?

File: torch/_inductor/kernel/bmm.py [hunk 1]
```diff
@@ -34,39 +34,52 @@ def bmm_grid(b, m, n, meta):
...
-    A = A + (ram[:, None] * stride_am + rk[None, :] * stride_ak + idx_q*stride_aq)
-    B = B + (rk[:, None] * stride_bk + rbn[None, :] * stride_bn + idx_q*stride_bq)
+
+    A_block_ptr = tl.make_block_ptr(
+        base=A + idx_q * stride_aq,
+        shape=(M, K),
+        strides=(stride_am, stride_ak),
+        offsets=(block_offset_m, 0),
+        block_shape=(BLOCK_M, BLOCK_K),
+        order=(1, 0),
```
Comment (jansel) [inline torch/_inductor/kernel/bmm.py:59]: `order=` should not be hardcoded

File: torch/_inductor/kernel/bmm.py [hunk 2]
```diff
@@ -34,39 +34,52 @@ def bmm_grid(b, m, n, meta):
...
+    )
+    B_block_ptr = tl.make_block_ptr(
+        base=B + idx_q * stride_bq,
+        shape=(K, N),
+        strides=(stride_bk, stride_bn),
+        offsets=(0, pid_n * BLOCK_N),
+        block_shape=(BLOCK_K, BLOCK_N),
+        order=(1, 0)
```
Comment (jansel) [inline torch/_inductor/kernel/bmm.py:67]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm.py [hunk 3]
```diff
@@ -44,39 +44,53 @@
...
+    block_offset_m = pid_m * BLOCK_M
+    block_offset_n = pid_n * BLOCK_N
+
+    A_block_ptr = tl.make_block_ptr(
+        base=A,
+        shape=(M, K),
+        strides=(stride_am, stride_ak),
+        offsets=(block_offset_m, 0),
+        block_shape=(BLOCK_M, BLOCK_K),
+        order=(1, 0),
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:67]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm.py [hunk 4]
```diff
@@ -44,39 +44,53 @@
...
+    )
+    B_block_ptr = tl.make_block_ptr(
+        base=B,
+        shape=(K, N),
+        strides=(stride_bk, stride_bn),
+        offsets=(0, pid_n * BLOCK_N),
+        block_shape=(BLOCK_K, BLOCK_N),
+        order=(1, 0)
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:75]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 5]
```diff
@@ -40,55 +40,82 @@
...
+    block_offset_m = pid_m * BLOCK_M
+    block_offset_n = pid_n * BLOCK_N
+
+    A_block_ptr = tl.make_block_ptr(
+        base=A,
+        shape=(M, K),
+        strides=(stride_am, stride_ak),
+        offsets=(block_offset_m, 0),
+        block_shape=(BLOCK_M, BLOCK_K),
+        order=(1, 0),
```
Comment (jansel) [inline torch/_inductor/kernel/mm_plus_mm.py:64]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 6]
```diff
@@ -40,55 +40,82 @@
...
+    )
+    B_block_ptr = tl.make_block_ptr(
+        base=B,
+        shape=(K, N),
+        strides=(stride_bk, stride_bn),
+        offsets=(0, pid_n * BLOCK_N),
+        block_shape=(BLOCK_K, BLOCK_N),
+        order=(1, 0)
```
Comment (jansel) [inline torch/_inductor/kernel/mm_plus_mm.py:72]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 7]
```diff
@@ -40,55 +40,82 @@
...
+    )
+    C_block_ptr = tl.make_block_ptr(
+        base=C,
+        shape=(M, K),
+        strides=(stride_cm, stride_ck),
+        offsets=(block_offset_m, 0),
+        block_shape=(BLOCK_M, BLOCK_K),
+        order=(1, 0),
```
Comment (jansel) [inline torch/_inductor/kernel/mm_plus_mm.py:80]: `order=` should not be hardcoded

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 8]
```diff
@@ -40,55 +40,82 @@
...
+    )
+    D_block_ptr = tl.make_block_ptr(
+        base=D,
+        shape=(K, N),
+        strides=(stride_dk, stride_dn),
+        offsets=(0, pid_n * BLOCK_N),
+        block_shape=(BLOCK_K, BLOCK_N),
+        order=(1, 0)
```
Comment (jansel) [inline torch/_inductor/kernel/mm_plus_mm.py:88]: `order=` should not be hardcoded

**Review by jansel — CHANGES_REQUESTED** [review]
I'm curious if adding more configs would help the perf on smaller sizes.

File: torch/_inductor/kernel/bmm.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_inductor/kernel/bmm.py:59]: Hmm, this depends on the layout. I guess I can check if stride_am > stride_ak?

Comment (jansel) [reply torch/_inductor/kernel/bmm.py]: You could add a stride_order helper similar to:
https://github.com/pytorch/pytorch/blob/13af952f94e611f67ac306233898b753a9cf73d2/torch/_inductor/select_algorithm.py#L205

Something like:
```
sym_strides = self.named_input_nodes[name].get_stride()
assert len(sym_strides) == 2
if V.graph.sizevars.evaluate_expr(sympy.Lt(sym_strides[0], sym_strides[1])):
  return "(0, 1)"
else:
  return "(1, 0)"
```

**Comment by OrenLeung** [comment]
@jon-chuang hi, any updates on this?

It seems that block ptr needed in order to make H100s efficient with its TMA.

**Pull request closed without merging**